### PR TITLE
[reland] use scatter in shard_tensor API

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #75991

This is a reland of PR https://github.com/pytorch/pytorch/pull/72160. The previous PR failed on some cases where un-even scatter happens. So this PR made some additional fixes to ensure it scatters correctly.
1. Fix a bug in ProcessGroupNCCL::scatter, which is a similar issue to https://github.com/pytorch/pytorch/pull/75535
2. resize the shard to the same size before calling `dist.scatter`, and resize it back to the original layout after receiving from scatter.

Differential Revision: [D35726920](https://our.internmc.facebook.com/intern/diff/D35726920/)

### CODE REVIEW

File: torch/distributed/_shard/sharding_spec/api.py [hunk 1]
```diff
@@ -195,31 +193,69 @@ def shard(self, tensor: torch.Tensor, src_rank: int = 0, process_group=None) ->
...
+        if list(local_tensor.size()) != local_metadata.shard_sizes:
+            # detach again after receiving to ensure local shards remain a leaf node
+            local_tensor = local_tensor.resize_(local_metadata.shard_sizes).detach()
+
+        # Sync requires_grad to local_shard.
+        local_tensor.requires_grad = tensor.requires_grad
+
+        local_shards.append(Shard(tensor=local_tensor, metadata=local_metadata))
+
+        st = ShardedTensor._init_from_local_shards_and_global_metadata(
```
Comment (fduwjj) [inline torch/distributed/_shard/sharding_spec/api.py:254]: This will also fix the metadata order we have discussed about before, right? I mean when the placements are like: 
```
placements=[
    "rank:2/cuda:2",
    "rank:3/cuda:3",
    "rank:0/cuda:0",
    "rank:1/cuda:1",
]
```

File: torch/distributed/_shard/sharding_spec/api.py [hunk 2]
```diff
@@ -195,31 +193,69 @@ def shard(self, tensor: torch.Tensor, src_rank: int = 0, process_group=None) ->
...
-                        metadata=shard_meta
-                    )
-                )
+                narrowed_tensor = narrow_tensor(tensor, shard_meta)
+                if shard_meta.shard_sizes[self.dim] < split_size:  # type: ignore[index]
+                    # for the last shard that might be smaller to other shards
+                    # resize the narrowed tensor to the same size and use it for
+                    # the scatter collective as dist.scatter requires same size
+                    # inputs on every rank
+                    tensor_to_scatter = narrowed_tensor.detach().clone().resize_(scatter_shape)
```
Comment (fduwjj) [inline torch/distributed/_shard/sharding_spec/api.py:221]: I tested it locally, this is essentially same as padding, random numbers are just added into the tensor for a bigger shape.. From the wiki of `resize_`, it says "Existing elements are preserved but any new memory is uninitialized.", I am not sure if there could be any side effect. If not, I think it's good to go.

**Review by fduwjj — COMMENTED** [review]
Overall, looks good to me but let's wait a little to see how others think of this solution. If no objection, I will stamp on it later this week. Also might be manually run some sharded op tests in case CI didn't capture potential failure. Thanks!

File: torch/distributed/_shard/sharding_spec/api.py [hunk 3]
```diff
@@ -195,31 +193,69 @@ def shard(self, tensor: torch.Tensor, src_rank: int = 0, process_group=None) ->
...
+                else:
+                    tensor_to_scatter = narrowed_tensor.detach().clone().contiguous()
+
+                tensors_to_scatter[rank] = tensor_to_scatter
+
+            if current_rank == rank:
```
Comment (pritamdamania87) [inline torch/distributed/_shard/sharding_spec/api.py:227]: nit: This is the same condition as above and can be merged?

Comment (pritamdamania87) [inline torch/distributed/_shard/sharding_spec/api.py:227]: nit: This is the same condition as above and can be merged?

**Review by pritamdamania87 — APPROVED** [review]

File: torch/distributed/_shard/sharding_spec/api.py (hunk 1 above)
Comment (wanchaol) [reply torch/distributed/_shard/sharding_spec/api.py:254]: Yep it should fix it

File: torch/distributed/_shard/sharding_spec/api.py (hunk 2 above)
Comment (wanchaol) [reply torch/distributed/_shard/sharding_spec/api.py:221]: Yeah I also looked into the `gather_object`  and `all_gather_object` in the c10d API, seems like they are all using `resize_` ([link](https://github.com/pytorch/pytorch/blob/b369b89f235f54bc9de85d768fb62ac4579681dc/torch/distributed/distributed_c10d.py#L1761)) So I think we are good here.

I am even thinking maybe we should switch the `pad` and `unpad` to also use `resize_` to consolidate our approaches on the non-uniform inputs.

File: torch/distributed/_shard/sharding_spec/api.py (hunk 3 above)
Comment (wanchaol) [reply torch/distributed/_shard/sharding_spec/api.py:227]: Hmm so it's a bit different, the first condition is `current_rank=src_rank` where it collects all the shards to scatter (only on src_rank), the second condition is `current_rank = rank` where it check if we are in the rank of the current `shard_metadata`.

File: torch/distributed/_shard/sharding_spec/api.py (hunk 2 above)
Comment (fduwjj) [reply torch/distributed/_shard/sharding_spec/api.py:221]: I see. Is there difference in padding and resize? I mean which one is more efficient?
